Add UUID version 7 as the default guid generator

ChrisJollyAU opened the pull request:
From discussion in #2909 

This adds a `GUID` generator compatible with the UID Version 7.

Code is based on the `Guid.CreateVersion7` function that is part of .Net 9. Since EFCore and EFCore.PG target .Net 8 we don't get that function by default

roji commented:
@ChrisJollyAU without going into the fine details, that looks good overall to me... @Timovzl can I ask you to sign off on this as well?

Timovzl commented on src/EFCore.PG/ValueGeneration/UUid7ValueGenerator.cs:
```diff
@@ -0,0 +1,59 @@
+///     Generates sequential <see cref="Guid" /> values according to the UUID version 7 specification.
+///     Will be updated to use Guid.CreateVersion7 when available.
+/// </summary>
+public class UUid7ValueGenerator : ValueGenerator<Guid>
```
**Suggestion:** The .NET team has chosen to stick to avoid using multiple synonyms, hence `Guid.CreateVersion7`. For consistency, I would go with `GuidVersion7ValueGenerator`.

Timovzl replied:
**Question:** All generators in the current package are prefixed with `Npgsql`. Should we follow that convention or not, @roji?

Timovzl replied:
Moreover, they are marked as internal APIs. Should we rather be consistent with them or with EF Core's own `GuidValueGenerator`?

Timovzl replied:
Let's have the prefix, for consistency with similar existing types in the package, and to reinforce the pubternal-ness some more.

Timovzl commented on src/EFCore.PG/ValueGeneration/UUid7ValueGenerator.cs:
```diff
@@ -0,0 +1,59 @@
+        Span<byte> guidBytes = stackalloc byte[16];
+        var succeeded = Guid.NewGuid().TryWriteBytes(guidBytes);
+        var unixms = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+        Span<byte> counterBytes = stackalloc byte[sizeof(long)];
```
**Issue:** The name `counterBytes` seems a bit misleading for the timestamp component, esp. now that we have no counter at all.

ChrisJollyAU replied:
Can easily change that. Besides if we use your suggestion it automatically falls away

ChrisJollyAU replied:
Different implementation used which doesn't have the above code, so resolved

Timovzl commented:
@ChrisJollyAU Below is my implementation using an almost-pure copy of the .NET 9 code. The only deviations are the `ref var resultClone` line and the references to `resultClone` instead of `result`.

While the downside is that this needs the extra struct, we get to use the source code as directly as possible. I feel safest staying as close to that as possible. Let me know your thoughts.

```cs
public class NpgsqlGuidVersion7ValueGenerator : ValueGenerator<Guid>
{
    public override bool GeneratesTemporaryValues => false;

    public override Guid Next(EntityEntry entry) => BorrowedFromNet9.CreateVersion7(timestamp: DateTimeOffset.UtcNow);

    // Code borrowed from .NET 9 should be removed as soon as the target framework includes such code
    #region Borrowed from .NET 9

#pragma warning disable IDE0007 // Use implicit type -- Avoid changes to code borrowed from BCL

    // https://github.com/dotnet/runtime/blob/f402418aaed508c1d77e41b942e3978675183bfc/src/libraries/System.Private.CoreLib/src/System/Guid.cs
    private static class BorrowedFromNet9
    {
        private const byte Variant10xxMask = 0xC0;
        private const byte Variant10xxValue = 0x80;

        private const ushort VersionMask = 0xF000;
        private const ushort Version7Value = 0x7000;

        /// <summary>Creates a new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</summary>
        /// <returns>A new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</returns>
        /// <remarks>
        ///     <para>This uses <see cref="DateTimeOffset.UtcNow" /> to determine the Unix Epoch timestamp source.</para>
        ///     <para>This seeds the rand_a and rand_b sub-fields with random data.</para>
        /// </remarks>
        public static Guid CreateVersion7() => CreateVersion7(DateTimeOffset.UtcNow);

        /// <summary>Creates a new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</summary>
        /// <param name="timestamp">The date time offset used to determine the Unix Epoch timestamp.</param>
        /// <returns>A new <see cref="Guid" /> according to RFC 9562, following the Version 7 format.</returns>
        /// <exception cref="ArgumentOutOfRangeException"><paramref name="timestamp" /> represents an offset prior to <see cref="DateTimeOffset.UnixEpoch" />.</exception>
        /// <remarks>
        ///     <para>This seeds the rand_a and rand_b sub-fields with random data.</para>
        /// </remarks>
        public static Guid CreateVersion7(DateTimeOffset timestamp)
        {
            // NewGuid uses CoCreateGuid on Windows and Interop.GetCryptographicallySecureRandomBytes on Unix to get
            // cryptographically-secure random bytes. We could use Interop.BCrypt.BCryptGenRandom to generate the random
            // bytes on Windows, as is done in RandomNumberGenerator, but that's measurably slower than using CoCreateGuid.
            // And while CoCreateGuid only generates 122 bits of randomness, the other 6 bits being for the version / variant
            // fields, this method also needs those bits to be non-random, so we can just use NewGuid for efficiency.
            Guid result = Guid.NewGuid();

            // 2^48 is roughly 8925.5 years, which from the Unix Epoch means we won't
            // overflow until around July of 10,895. So there isn't any need to handle
            // it given that DateTimeOffset.MaxValue is December 31, 9999. However, we
            // can't represent timestamps prior to the Unix Epoch since UUIDv7 explicitly
            // stores a 48-bit unsigned value, so we do need to throw if one is passed in.

            long unix_ts_ms = timestamp.ToUnixTimeMilliseconds();
            ArgumentOutOfRangeException.ThrowIfNegative(unix_ts_ms, nameof(timestamp));

            ref var resultClone = ref Unsafe.As<Guid, GuidDoppleganger>(ref result); // Deviation from BLC: Reinterpret Guid as our own type so that we can manipulate its private fields

            Unsafe.AsRef(in resultClone._a) = (int)(unix_ts_ms >> 16);
            Unsafe.AsRef(in resultClone._b) = (short)(unix_ts_ms);

            Unsafe.AsRef(in resultClone._c) = (short)((resultClone._c & ~VersionMask) | Version7Value);
            Unsafe.AsRef(in resultClone._d) = (byte)((resultClone._d & ~Variant10xxMask) | Variant10xxValue);

            return result;
        }
    }

    /// <summary>
    /// Used to manipulate the private fields of a <see cref="Guid"/> like its internal methods do, by treating a <see cref="Guid"/> as a <see cref="GuidDoppleganger"/>.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    internal readonly struct GuidDoppleganger
    {
#pragma warning disable IDE1006 // Naming Styles -- Avoid further changes to code borrowed from BCL when working with the current type
        internal readonly int _a;   // Do not rename (binary serialization)
        internal readonly short _b; // Do not rename (binary serialization)
        internal readonly short _c; // Do not rename (binary serialization)
        internal readonly byte _d;  // Do not rename (binary serialization)
        internal readonly byte _e;  // Do not rename (binary serialization)
        internal readonly byte _f;  // Do not rename (binary serialization)
        internal readonly byte _g;  // Do not rename (binary serialization)
        internal readonly byte _h;  // Do not rename (binary serialization)
        internal readonly byte _i;  // Do not rename (binary serialization)
        internal readonly byte _j;  // Do not rename (binary serialization)
        internal readonly byte _k;  // Do not rename (binary serialization)
#pragma warning restore IDE1006 // Naming Styles
    }

#pragma warning restore IDE0007 // Use implicit type

    #endregion
}
```

Timovzl commented on src/EFCore.PG/ValueGeneration/Internal/NpgsqlValueGeneratorSelector.cs:
```diff
@@ -104,6 +104,6 @@ public override bool TrySelect(IProperty property, ITypeBase typeBase, out Value
             ? property.ValueGenerated == ValueGenerated.Never || property.GetDefaultValueSql() is not null
                 ? new TemporaryGuidValueGenerator()
-                : new GuidValueGenerator()
+                : new UUid7ValueGenerator()
```
We should have some tests on this line as well as on the generator's implementation. I think we should at least include:

- UUIDv7 should be generated by default.
- The version and variant bits should be set as expected.
- The timestamp bits should reflect the timestamp (at least coarsely asserted).
- The final bits should have a high entropy.

roji replied:
IMHO there's a bit of diminishing returns on this kind of testing - the test basically ends up duplicating the implementation. I do agree it makes sense to make sure that the generator is actually being used by default (e.g. that the version/variant bits are correct), but beyond that I personally wouldn't invest too more effort in e.g. asserting on the timestamp.

roji replied:
Not to mention that this is all supposed to go away in the next release...

ChrisJollyAU replied:
I already have a test in EFCore.Jet that I can copy to here. Just need to know where to put it

Timovzl replied:
Fair. I did not mean to imply that we test the UUIDv7 implementation's inner workings, but mainly assert the fact _that_ UUIDv7 is being used. I agree that the version and variant are the primary indicators.

roji commented:
Thanks for the additional implementation proposal @Timovzl. I don't have strong feels between your doppleganger reinterpret cast solution and @ChrisJollyAU - remember this is all planned to go away in a year, so the important thing is to have something that works here. @ChrisJollyAU any comment?

ChrisJollyAU commented:
Don't mind whichever way. I'm happy to adjust to @Timovzl suggestion. Practically its modifying the data as the int and short as part of a struct versus the individual bytes. Might end up a couple less instructions on the struct but probably won't make too much noticeable difference

Timovzl commented:
In the absence of thorough testing, my preference would be to stay as close as possible to the source implementation.

Whichever we pick, perhaps we can include a Theory that, for a handful of timestamps, asserts that the result is identical to the output of `Guid.CreateVersion7(timestamp)` (as a constant passed to the theory, obtained from .NET 9 by hand).

Timovzl commented:
@ChrisJollyAU Nice. Looks like it needs .NET 9, though, so we might need to precalculate the `net9internal` value and parse that from a constant here, right?

ChrisJollyAU commented:
@Timovzl I must check how efcore.pg does the tests. I know my EFCore.Jet setup has the provider targeting .Net 8 but the tests target .Net 9

roji commented:
For testing, check out NpgsqlValueGeneratorSelectorTest - this is a unit test suite that would be the easiest/ideal place to place such a test.

Given we're all OK with both implementation options, I'd say let's go with @Timovzl suggestion as the closest to the .NET implementation (but once again am fine either way). Once the PR is updated and some minimal testing is added I'm happy to merge that for 9.0!

ChrisJollyAU commented:
@roji Thanks. Added the test there. As it happens your test project does target .Net 9 (preview 3) so we can use `Guid.CreateVersion7`

I've moved the value generator to the Internal folder under ValueGeneration so that it gets the `.Internal` added to its namespace

@Timovzl I needed to access the `BorrowedFromNet9` from the test so I couldn't use it as `private`. Made it `internal` instead

Timovzl commented on src/EFCore.PG/ValueGeneration/Internal/NpgsqlUUID7ValueGenerator.cs:
```diff
@@ -0,0 +1,108 @@
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.ValueGeneration.Internal;
+
+/// <summary>
```
**Chore:** Let's replace this summary, and the two below it, by the "pubternal" summary (copied from similar types elsewhere in the project):

```cs
/// <summary>
///     This API supports the Entity Framework Core infrastructure and is not intended to be used
///     directly from your code. This API may change or be removed in future releases.
/// </summary>
```

ChrisJollyAU replied:
Everything else in that ValueGeneration folder looks like it uses that wording so maybe keep it like that to not be different to other value generators

roji replied:
It's true that the EF standard is to have the standard "pubternal" summary on everything - both the class and anything public/protected on it. This probably isn't applied consistently in the provider at the moment, and also isn't extremely important.

ChrisJollyAU replied:
Updated summary description

Timovzl commented on test/EFCore.PG.Tests/NpgsqlValueGeneratorSelectorTest.cs:
```diff
@@ -210,4 +210,19 @@ public override int Next(EntityEntry entry)
     }
+
+    [Fact]
+    public void CustomUuid7Test()
```
**Nitpick:** Based on the naming of other tests here, I'd suggest `NpgsqlUuid7ValueGenerator_creates_uuidv7()`. In fact, that test probably belongs in a new test class specific to that class, in which case the name can just be `Creates_uuidv7()`.

ChrisJollyAU replied:
Good suggestion for name. Might want to just keep it in here. Reason being it is only temporary. Do we really want to create a new class, just for something temporary?

Timovzl replied:
Agreed, let's keep it here.

Timovzl replied:
Hm... Isn't the new generator here to stay? And only its implementation is going to be much simplified? It ought to have a test that confirms that it does what it claims, even when what it does is piggyback on .NET 9, right?

ChrisJollyAU replied:
Wouldn't that be doing the same thing as the other test mentioned further down then?

Timovzl commented on src/EFCore.PG/ValueGeneration/Internal/NpgsqlUUID7ValueGenerator.cs:
```diff
@@ -0,0 +1,108 @@
+///     Generates sequential <see cref="Guid" /> values according to the UUID version 7 specification.
+///     Will be updated to use Guid.CreateVersion7 when available.
+/// </summary>
+public class NpgsqlUUid7ValueGenerator : ValueGenerator<Guid>
```
**Typo:** The class name says `UUid`, which should be `Uuid`. The file name says `UUID`, which should also be `Uuid`.

ChrisJollyAU replied:
Name of class changed, but filename change only worked locally

Timovzl replied:
Ah yeah, that annoying issue. Add a digit or something, `*2.cs`, push, then rename to the desired casing and push again.

Timovzl replied:
Edit: If we add the Npgsql prefix, the casing change would also work.

Timovzl replied:
Hm, did you add the prefix without adjusting the casing, by any chance?

ChrisJollyAU replied:
File name should be sorted now

Timovzl commented on src/EFCore.PG/ValueGeneration/Internal/NpgsqlValueGeneratorSelector.cs:
```diff
@@ -104,6 +104,6 @@ public override bool TrySelect(IProperty property, ITypeBase typeBase, out Value
             ? property.ValueGenerated == ValueGenerated.Never || property.GetDefaultValueSql() is not null
                 ? new TemporaryGuidValueGenerator()
-                : new GuidValueGenerator()
+                : new NpgsqlUUid7ValueGenerator()
```
**Suggestion:** I think we should also have a test that confirms that this is now the standard behavior when working with a `Guid` that allows value generation. That particular test could simply check the version indicator, as far as I'm concerned.

roji replied:
The current NpgsqlUuid7ValueGenerator_creates_uuidv7 test as it is in the PR tests the generator, but I think it should be testing NpgsqlValueGeneratorSelector.cs, as the test class implies and as other tests do. In other words, like in the other tests, the test can set up a model with a Guid key property, then get a value generated for that property and assert that it's what we want. That would confirm that it's the standard behavior.

roji replied:
Having said all that, the combination of the two tests should indeed be sufficient - above you make sure that your generator is used by default, and below you make sure that that it produces the right kind of Guids.

@Timovzl or do you still see a problem here?

Timovzl replied:
Agreed, either way would assert the desired combination of factors.

@ChrisJollyAU Could you either add a test that confirms UUIDv7 as the default for a `Guid`, or adjust the existing test to test the selector for a `Guid` rather than a direct invocation? (If in doubt, I'd lean towards the former, to separate responsibilities: `Guid` should get a UUIDv7 by default; and any custom implementation we use should behave as expected when it comes to the details.)

roji replied:
See my comments above about the current two tests probably being sufficient. Not against more testing, but not sure it adds value.

Timovzl replied:
@roji, @ChrisJollyAU, sorry, I misunderstood. Saw only one new test, but hadn't realized that we were already asserting that `Guid` was handled by the new generator: `AssertGenerator<NpgsqlUuid7ValueGenerator>("Guid");`

That resolves this comment. :)

ChrisJollyAU replied:
Just ran through the tests locally and found another place that uses the `AssertGenerator` which is now fixed

roji commented:
@ChrisJollyAU is this ready for a final review?

ChrisJollyAU commented:
@roji Should be. Unless there's anything I've missed.

CI doesn't seem to be passing but if it matches what I have locally it is unrelated to this PR. (A compile time syntax error in test `String_Join_non_aggregate` and some errors in the DatabaseModelFactoryTest `42809: "BlogsView" is not a table`)

roji reviewed (approved):
Thanks for the work here @ChrisJollyAU and @Timovzl - much appreciated!

Compilation errors were because the project was still using dotnet SDK 9.0.0-preview.3, which did not yet contain the new Guid.CreateVersion7() API - I pushed a commit to bump that to preview.7. I also commented out a test that's causing compilation issues, I believe the problem is already solved in later SDKs and will make sure to address before the release.

Other than that everything looks great, thanks!
